### PR TITLE
Adding Copy and Move Constructors for RNN and FastLSTM layer.

### PR DESCRIPTION
**Pull request by Abilityguy**
Reference #2625 
This PR attempts to add the copy and move constructors to RNN and FastLSTM layers.

### CODE REVIEW

**Comment by Abilityguy** [comment]
Hi @himanshupathak21061998 , I tried to add the copy and move constructors to FastLSTM layer. 

However, to write a test case for the same similar to the ones implemented for FFN (like the one [here)](https://github.com/mlpack/mlpack/pull/2711/files#diff-338dcffcd150ae0b9b5c3c700d8aac8dbe9bd125a3f411d4fbccc2dd99a41be6), I believe copy and move constructors need to be added to the RNN implementation. I tried to write one referring to your PR #2324 . However, I am getting errors when I test this on my local system. Could you take a look at the changes and point out where I went wrong?

File: src/mlpack/tests/ann_layer_test.cpp [hunk 1]
```diff
@@ -1183,6 +1183,68 @@ TEST_CASE("FastLSTMLayerParametersTest", "[ANNLayerTest]")
...
+ model1->Add<FastLSTM<> >(10, 3, rho);
+ model1->Add<LogSoftMax<> >();
+
+ ens::StandardSGD opt(0.1, 1, 5, -100, false);
+ model1->Train(input, target, opt);
+
+ arma::cube predictions1;
+ model1->Predict(input, predictions1);
+
+ RNN<> model2() = *model1;
```
Comment (himanshupathak21061998) [inline src/mlpack/tests/ann_layer_test.cpp:1209]: Hey @Abilityguy thanks for contributing.
Just remove `()` from here we don't need them to call copy constructor.
```
RNN<> model2 = *model1;
```

**Comment by Abilityguy** [comment]
Thanks for the quick review @himanshupathak21061998 .
Your fix worked and the test cases passed. 

However, I then tried to refactor the code into functions ```CheckRNNCopyFunction``` and ```CheckRNNMoveFunction``` so that it would be easier to test the other copy and move layers. 
Now I get a ```std::bad_alloc``` error for the ```CheckCopyMoveFastLSTMTest``` test case.

I think the error is in the ```CheckRNNCopyFunction``` since commenting out that function call makes the test cases pass. I have not been able to figure out the reason for this error.

File: src/mlpack/tests/ann_layer_test.cpp [hunk 2]
```diff
@@ -31,6 +31,52 @@
...
+  ens::StandardSGD opt(0.1, 1, maxEpochs * trainData.n_cols, -100, false);
+
+  network1->Train(trainData, trainLabels, opt);
+  network1->Predict(trainData, predictions1);
+
+  RNN<> network2 = *network1;
+  delete network1;
+
+  // Deallocating all of network1's memory, so that
+  // if network2 is trying to use any of that memory.
```
Comment (rcurtin) [inline src/mlpack/tests/ann_layer_test.cpp:52]: Do you want to fix this sentence?  There is no `then` with the `if`. :smile:

File: src/mlpack/methods/ann/rnn.hpp [hunk 3]
```diff
@@ -70,6 +70,15 @@ class RNN
...
       InitializationRuleType initializeRule = InitializationRuleType());
 
+  //! Copy constructor.
+  RNN(const RNN&);
+
+  //! Move constructor.
+  RNN(RNN&&);
+
+  //! Copy/move assignment operator.
+  RNN& operator = (RNN);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/rnn.hpp:80]: Maybe I overlooked it, but I don't see an implementation of this function.

**Review by rcurtin — COMMENTED** [review]
Hey @Abilityguy, thanks for working on this!  The implementation looks fine so I am a little surprised that you are getting a failure.  If you are getting a `error: arma::memory::acquire(): out of memory`, this generally means that sometimes a matrix of way too large a size is being allocated.  You should try using a debugger to catch when the exception is thrown, and then see what the size of the matrix was that the code was attempting to allocate.  You could use `gdb` for this, for instance.

File: src/mlpack/tests/ann_layer_test.cpp (hunk 2 above)
Comment (Abilityguy) [reply src/mlpack/tests/ann_layer_test.cpp]: Done. Fixed it in other files too 👍🏼 .

File: src/mlpack/methods/ann/rnn.hpp (hunk 3 above)
Comment (Abilityguy) [reply src/mlpack/methods/ann/rnn.hpp]: Was trying to follow the convention used in ```mlpack/methods/ann/ffn.hpp``` but now have changed it to be clearer. Both the copy and move implementations are in ```rnn_impl.hpp```.

**Comment by Abilityguy** [comment]
Thanks for the suggestion @rcurtin !. 
So I played around with ```gdb``` and found the error. 

The problem was with the ```FastLSTM``` copy implementation. Specifically, the ```batchSize``` parameter was not copied in the copy implementation, which made ```network2``` use the ```batchSize``` value in memory (usually some huge garbage value). When ```network2``` used this value in prediction, it requested for a huge matrix which causes the ```std::bad_alloc``` error.

Made the fix by including ```batchSize``` and other variables to the copy and move implementations.

**Comment by rcurtin** [comment]
Awesome work @Abilityguy!  gdb can be really helpful for digging to the bottom of what's going on. I use it far more often than I would like to... 😄

File: src/mlpack/methods/ann/rnn_impl.hpp [hunk 4]
```diff
@@ -49,6 +49,49 @@ RNN<OutputLayerType, InitializationRuleType, CustomLayers...>::RNN(
...
+    inputSize(std::move(network.inputSize)),
+    outputSize(std::move(network.outputSize)),
+    targetSize(std::move(network.targetSize)),
+    reset(std::move(network.reset)),
+    single(std::move(network.single)),
+    parameter(std::move(network.parameter)),
+    numFunctions(std::move(network.numFunctions)),
+    deterministic(std::move(network.deterministic))
+{
+  this->network = std::move(network.network);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/rnn_impl.hpp:92]: You could put this in the initialization list too if you liked. :+1:

**Review by rcurtin — APPROVED** [review]
Thanks @Abilityguy!  Looks good to me.  Nice work solving the issue. :100:

File: src/mlpack/methods/ann/rnn_impl.hpp (hunk 4 above)
Comment (Abilityguy) [reply src/mlpack/methods/ann/rnn_impl.hpp]: Done.

**Comment by Abilityguy** [comment]
> Awesome work @Abilityguy! gdb can be really helpful for digging to the bottom of what's going on. I use it far more often than I would like to...

Right!. I am gonna start using it more now. 😄

**Comment by rcurtin** [comment]
Thabks @Abilityguy!

**Pull request merged**